### ISSUE TITLE
Folder structuring.

### ISSUE THREAD

**Issue by mindrage** [otland/forgottenserver#1057]
I find the data folder really useful when creating scripts, its as simple as selecting the event you want to make, then make the script for it. However. It also is something i think most people find annoying. Everything is clustered around all diffrent folders, you dont really have much options to organise.
Yes, you could link the folders with the ".." paths but. nothing really anybody prefers to work with.

So.. i come with this suggestion instead. 
What if we made the data folders modular. Keep the same structure inside the data folder. But allow multiple data folders to be loaded in. and only load in those folders which exists inside a folder.

The folders will be linked via a xml file, which has these attributes:  <Module path="Quests/Annihilator" priority="1" enabled="yes/no"/>

this would basically make systems much more organised and work can be shared and edited separately.
The current data folder could be called core, main, default or something.

So the general benefits:
  *Grouping scripts, allowing people to work on a system and then only need to paste a folder to combine systems.
  *Ability to disable systems
  *Little to low coding, You practically only need to iterate the folders in the modules.xml file, and maybe make folders optional for everything except the main module.
  *Backwards compability, Everything you written should work out of the box.

The drawbacks would be small. 
  *a bit longer loading time, since obviously more files to load.
  *The modules obviously would contains alot of folders, ex Talkaction/scripts/ will still be there for every folder that uses a talkaction..

Since i've heard many people also wanted mods back into the code, this would actually be a simple but pretty flexible solution.
##

**Comment by Zbizu** [comment]
modules.lua could look like this:
````
registerAction(uniqueId, 30015, "--[[data/modules/]] quests/annihilator.lua")
registerMonster("Troll", "Trolls/troll.xml")
````

everything can be loaded from global.lua anyway so if you want to get rid of xml files, it may become main script loader, for example:
global.lua
````
dofile('data/actions/actions.lua')
````
actions.lua would look like this:
````
registerAction(uniqueId, 30015, "--[[data/actions/]] quests/annihilator.lua")
````

**Comment by slawkens** [comment]
Basically what you mean is mods. They're already included in TFS 0.3/0.4. You can make multiple systems. The idea is simple: you have one file per system. In that file you can define as many actions/talkactions/items etc as you wish.

I might make support for mods for TFS 1.0 when more people are interested.

**Comment by mindrage** [comment]
Yeah, the idea was just to be able to package things together for easier management and (maybe) distrobution. Having it in lua would be pretty great. but security is also a factor.

**Comment by slawkens** [comment]
The best solution would be probably adding simple lua functions for registering actions/items etc.

Like this:
```
function myFunction(cid, ...)
	print('Test')
end

Actions:registerActionId(300, myFunction)
Movements:registerUniqueId(100, myFunction)

local properties = {weight = 100, value = 1}
Items:register(2160, "gold coin", properties)

Talkactions:register("!mycommand", myFunction)
```

And then, make a directory (Like plugins/mods) where are .lua files will be automatically loaded from.

**Comment by ranisalt** [comment]
We are talking about an entire module restructuring, not just folders, and that probably will come after a Lua interface restructuring.

**Comment by joseluis2g** [comment]
Some days ago @TheSumm posted this on Discord
https://github.com/TheSumm/forgottenserver/commits/action-script
That's pretty clean imo

**Comment by slawkens** [comment]
Better system has been developed round 10 YEARS ago.

That system is called MODS and is in TFS 0.3 series since ages.

Link if someone's is interested: https://github.com/slawkens/tfs-mods

**Comment by CristoferMartins** [comment]
@slawkens https://github.com/HeavenIsLost/elysium/blob/master/data/actions/actions.lua
Check this, i appreciate your feedback.

**Comment by ranisalt** [comment]
Hierarchical modules are nice, but not code inside XML. I would prefer something scripted in Lua instead.

**Comment by slawkens** [comment]
@ranisalt I also don't like the code inside of xml files, because you can't highlight it without plugins etc.. The examples I posted were just to demonstrate the power of the system. Normally it's possible to register them (any actions, events) in separate files. That's possible with the mods system. I just wanted to show that idea isn't new, and some work in this direction has been already done.

**Comment by Tofame** [comment]
This should be closed, as this is what revscriptsystem does and what people dicussed here is achieved with them.
